README.md
# pwcheckd: authenticate users sending their password over a UNIX socket

Used by sys-auth/pwcheckd from CLIP OS.


CHANGELOG
# CLIP ChangeLog for sys-auth/pwcheckd

*pwcheckd-1.1.7 (05 Sep 2014)

  05 Sep 2014; Tony Cheneau <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.7.ebuild, -pwcheckd-1.1.6.ebuild:
  Fix currently logged user detection. A change in last command behavior
  prompted a fix.

*pwcheckd-1.1.6 (07 Aug 2014)

  07 Aug 2014; Yves-Alexis Perez <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.6.ebuild, -pwcheckd-1.1.5-r1.ebuild:
  Fix implicit size_t -> int cast on amd64.

*pwcheckd-1.1.5-r1 (06 Jan 2011)
  06 Jan 2011; Benjamin Morin <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.5-r1.ebuild, -pwcheckd-1.1.5.ebuild:
  Removed paxctl privileges (became irrelevant since the PKCS11
  library is the client of the p11 proxy).

*pwcheckd-1.1.5 (09 Nov 2010)

  09 Nov 2010; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.5.ebuild, -pwcheckd-1.1.4-r1.ebuild:
  Minor fix to XAUTHORITY path in pwcheck_x11 (unused AFAIK).

*pwcheckd-1.1.4-r1 (08 Nov 2010)
  13 jul 2010; Benjamin Morin <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.4-r1.ebuild, -pwcheckd-1.1.4.ebuild:
  Modified the pam configuration (added pam_wheel module to
  choose the correct authentication method depending on the
  user belonging to pkauth group.

*pwcheckd-1.1.4 (13 Jul 2010)

  13 jul 2010; Benjamin Morin <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.4.ebuild, -pwcheckd-1.1.3.ebuild:
  Modified pwcheckd pam configuration to support smartcard
  authentication; modified the default timeout value
  (PW_TIMEOUT_DEFAULT) in checker.c.

*pwcheckd-1.1.3 (18 Mar 2009)

  18 Mar 2009; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.3.ebuild, -pwcheckd-1.1.2.ebuild:
  Fix gcc-4 compilation.

*pwcheckd-1.1.2 (26 Dec 2008)

  26 Dec 2008; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.2.ebuild, -pwcheckd-1.1.1.ebuild:
  Add support for X11 authentication socket type.

*pwcheckd-1.1.1 (23 Dec 2008)

  23 Dec 2008; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.1.ebuild, -pwcheckd-1.1.0.ebuild:
  Fix zombie childs.

*pwcheckd-1.1.0 (22 Dec 2008)

  22 Dec 2008; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.1.0.ebuild, -pwcheckd-1.0.1.ebuild:
  New code, supports typed sockets, multiple sockets and multiple
  connections.

*pwcheckd-1.0.1 (02 Apr 2008)

  02 Apr 2008; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.0.1.ebuild, -pwcheckd-1.0.ebuild:
  Add proper journalisation for PAM error codes.

*pwcheckd-1.0 (12 Oct 2007)

  12 Oct 2007; Vincent Strubel <clipos@ssi.gouv.fr>
  +pwcheckd-1.0.ebuild, +files/pwcheckd.pam:
  Initial import, brand new jailmaster lookalike:)


test/checker.c
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
/* Nyark nyark */

#include "../src/checker.c"

int 
main(int argc, char *argv[])
{
	struct auth_info auth;
	int ret;
	uid_t uid;

	if (argc < 4) {
		puts("Usage : checker <username> <passwd> <uid>");
		return EXIT_FAILURE;
	}
	
	auth.username = argv[1];
	auth.passwd = argv[2];
	uid = atoi(argv[3]);

	if (setuid(uid)) {
		perror("setuid");
		return EXIT_FAILURE;
	}

	ret = do_auth(&auth);

	if (ret)
		puts("failure");
	else
		puts("success");
	return ret;
}


test/request.c
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>

static int usage (const char *);

static int usage(const char *s)
{
	fprintf(stderr, "Usage: %s <path to socket> <passwd>\n", s);
	return 1;
}

int main(int argc, const char *argv[])
{
	int s;
	char c;
	struct sockaddr_un sau;
	ssize_t len;

	if (argc != 3)
		exit(usage(argv[0]));

	sau.sun_family = AF_UNIX;
	snprintf(sau.sun_path, sizeof(sau.sun_path), 
		"%s", argv[1]);

	s = socket(PF_UNIX, SOCK_STREAM, 0);
	if (s < 0) {
		perror("socket");
		return 1;
	}
	
	if (connect(s, (struct sockaddr *)&sau, sizeof(struct sockaddr_un)) < 0) {
		perror("connect");
		close(s);
		return 1;
	}
	
	len = write(s, argv[2], strlen(argv[2]));
	if (len < 0) {
		perror("write");
		close(s);
		return 1;
	}

	if (len != strlen(argv[2])) {
		puts("could not write all");
	}
	/* Wait for nested X clients to finish before waking xinit... */
	read(s, &c, 1);
	close(s);
	if (c == 'Y') {
		puts("yeepee");
		return EXIT_SUCCESS;
	} else {
		puts("bouh");
		return EXIT_FAILURE;
	}
}


test/timeout.c
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>

static int usage (const char *);

static int usage(const char *s)
{
	fprintf(stderr, "Usage: %s <path to socket>\n", s);
	return 1;
}

int main(int argc, const char *argv[])
{
	int s;
	char c;
	struct sockaddr_un sau;
	ssize_t len;

	if (argc != 2)
		exit(usage(argv[0]));

	sau.sun_family = AF_UNIX;
	snprintf(sau.sun_path, sizeof(sau.sun_path), 
		"%s", argv[1]);

	s = socket(PF_UNIX, SOCK_STREAM, 0);
	if (s < 0) {
		perror("socket");
		return 1;
	}
	
	if (connect(s, (struct sockaddr *)&sau, sizeof(struct sockaddr_un)) < 0) {
		perror("connect");
		close(s);
		return 1;
	}
	
	for (;;)
		sleep(1);
}


src/protos.h
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
/* 
 *  protos.h - pwcheckd prototypes
 *  Copyright (C) 2007 SGDN
 *  Author: Vincent Strubel <clipos@ssi.gouv.fr>
 *
 *  All rights reserved.
 *
 */


#ifndef _PROTOS_H
#define _PROTOS_H

#define _GNU_SOURCE
#ifdef USE_SYSLOG
#include <syslog.h>
#endif
#include <errno.h>
#include <string.h>
#include <stdio.h>

#define PREFIX_MASTER	"master"

extern const char *g_prefix;

#define _WARN(fmt, args...) fprintf(stderr, "[%s] "fmt"\n", g_prefix, ##args)
#define _WARN_ERRNO(fmt, args...) \
	fprintf(stderr, "[%s] "fmt": %s\n", g_prefix, ##args, strerror(errno))
#define _LOG(fmt, args...) printf("[%s] "fmt"\n", g_prefix, ##args)
#define _DEBUG(fmt, args...) printf("[%s](%s) "fmt"\n",g_prefix, \
						__FUNCTION__, ##args)

#ifdef USE_SYSLOG

#define WARN(fmt, args...) do { \
	if (g_daemonized) \
		syslog(LOG_WARNING, "[%s] "fmt"\n", g_prefix, ##args); \
	else \
		_WARN(fmt, ##args); \
} while (0)

#define WARN_ERRNO(fmt, args...) do {\
	if (g_daemonized) \
		syslog(LOG_WARNING, "[%s] "fmt": %s\n", g_prefix, \
						##args, strerror(errno)); \
	else \
		_WARN_ERRNO(fmt, ##args); \
} while (0)

#define WARN_PAM(fmt, pamh, errnum, args...) do {\
	if (g_daemonized) \
		syslog(LOG_WARNING, "[%s] "fmt": pam error: %s\n", g_prefix, \
					##args, pam_strerror(pamh, errnum)); \
	else \
		_WARN(fmt ": pam error: %s", ##args, pam_strerror(pamh, errnum)); \
} while (0)

#define LOG(fmt, args...) do {\
	if (g_daemonized) \
		syslog(LOG_INFO, "[%s] "fmt"\n", g_prefix, ##args); \
	else \
		_LOG(fmt, ##args); \
} while (0)

#define DEBUG(fmt, args...) do {\
	if (g_verbose) {\
		if (g_daemonized) \
			syslog(LOG_DEBUG, "[%s](%s) "fmt"\n", g_prefix, \
							__FUNCTION__, ##args); \
		else \
			_DEBUG(fmt, ##args); \
	} \
} while (0)

#else /* ! USE_SYSLOG */

#define WARN _WARN

#define WARN_PAM(fmt, pamh, errnum, args...) \
		_WARN(fmt ": pam error: %s", ##args, pam_strerror(pamh, errnum))

#define WARN_ERRNO _WARN_ERRNO

#define LOG _LOG

#define DEBUG(fmt, args...) do {\
	if (g_verbose)\
		_DEBUG(fmt, ##args);\
} while (0)

#endif /* ! USE_SYSLOG */

int check_user_self(int sock, char *sockname);
int check_user_x11(int sock, char *sockname);

extern int g_daemonized;
extern int g_verbose;

#define __U __attribute__((unused))

#endif /* _PROTOS_H */


src/checker.c
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
/* 
 *  checker.c - pwcheckd authentication functions
 *  Copyright (C) 2007 SGDN
 *  Author: Vincent Strubel <clipos@ssi.gouv.fr>
 *
 *  All rights reserved.
 *
 */


#define _GNU_SOURCE
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <pwd.h>
#include <grp.h>
#include <signal.h>
#include <sys/wait.h>

#include <security/pam_appl.h>
#include <clip/clip.h>

#include "protos.h"

#define PW_MAXLEN 64
#define PW_TRIES  1

#define PW_TIMEOUT_DEFAULT	25
#ifndef PW_TIMEOUT
#define PW_TIMEOUT	PW_TIMEOUT_DEFAULT
#endif

#define X11_TIMEOUT_DEFAULT	60
#ifndef X11_TIMEOUT
#define X11_TIMEOUT	X11_TIMEOUT_DEFAULT
#endif

#define PW_READ_TIMEOUT (PW_TIMEOUT * 500) /* msecs */

#define _TO_STR(var) #var
#define TO_STR(var) _TO_STR(var)

#define X11_SCRIPT TO_STR(SBINDIR)"/pwcheck_x11"
#define X11_SEP "\r\n"

struct auth_info {
	char *username;
	char *passwd;
};

#ifdef USE_TCB

#ifdef TCB_GROUP
#define _TCB_GROUP TO_STR(TCB_GROUP)
#else
#define _TCB_GROUP "shadow"
#endif 

static int
set_gid(void)
{
	const struct group *grp = getgrnam(_TCB_GROUP);
	if (!grp) {
		WARN_ERRNO("could not get gid for group %s", _TCB_GROUP);
		return -1;
	}

	if (setgid(grp->gr_gid)) {
		WARN_ERRNO("could not set gid to %d(%s)", 
					grp->gr_gid, _TCB_GROUP);
		return -1;
	}

	return 0;
}

#endif /* ! USE_TCB */


/*************************************************************/
/*                   Common PAM part                         */
/*************************************************************/

static int
conversation(int num_msg, const struct pam_message **msg, 
		struct pam_response **resp, void *data)
{
	int i;
	struct pam_response *reply;
	char *pwd;
	
	const struct auth_info *auth = data;

	reply = malloc(num_msg * sizeof(*reply));
	if (!reply)
		return PAM_CONV_ERR;

	for (i = 0; i < num_msg; i++) 
		reply[i].resp = NULL;
	
	for (i = 0; i < num_msg; i++) {
	    switch (msg[i]->msg_style) {
	        /* Ignore those */
	        case PAM_TEXT_INFO:
	        	reply[i].resp_retcode = PAM_SUCCESS;
	        	break;
	        case PAM_PROMPT_ECHO_OFF:
	        	pwd = strdup(auth->passwd);
	        	if (!pwd)
	        		goto out_free;
	        	reply[i].resp = pwd;
	        	reply[i].resp_retcode = PAM_SUCCESS;
	        	break;
	        /* We don't expect this, since we already gave the username
	         * to pam, but let's play it safe... */
	        case PAM_PROMPT_ECHO_ON:
	        	pwd = strdup(auth->username);
	        	if (!pwd)
	        		goto out_free;
	        	reply[i].resp = pwd;
	        	reply[i].resp_retcode = PAM_SUCCESS;
	        	break;
	        default:
	        	goto out_free;
	    }
	}
		
	*resp = reply;
	return PAM_SUCCESS;
			
out_free:
	for (i = 0; i < num_msg; i++) {
		if (reply[i].resp)
			free(reply[i].resp);
	}
	free(reply);
	return PAM_CONV_ERR;
}

static int
do_auth(struct auth_info *auth)
{
	int error;
	pam_handle_t *pamh;

	int ret = -1;
	struct pam_conv conv = {
		.conv = conversation,
		.appdata_ptr = auth,
	};

	error = pam_start("pwcheckd", auth->username, &conv, &pamh);
	if (error != PAM_SUCCESS) {
		WARN_PAM("could not start PAM", pamh, error);
		goto end_pam;
	}

	error = pam_authenticate(pamh, 0);
	if (error != PAM_SUCCESS) {
		WARN_PAM("could not authenticate", pamh, error);
		goto end_pam;
	}

	error = pam_acct_mgmt(pamh, 0);
	if (error != PAM_SUCCESS) {
		WARN_PAM("account management error", pamh, error);
		goto end_pam;
	}

	ret = 0;
	
end_pam:
	error = pam_end(pamh, error);
	if (error != PAM_SUCCESS) {
		WARN_PAM("could not end pam session", pamh, error);
		return -1;
	}

	return ret;
}

/*************************************************************/
/*                   Signal handling                         */
/*************************************************************/

static void 
sigalrm_action(int signum, siginfo_t *info, void *ctx __U)
{
	if (signum != SIGALRM) {
		WARN("Unexpected signal %d, expected SIGALRM", signum);
		return;
	}
	if (info->si_code != SI_KERNEL) {
		WARN("Received SIGALRM from unexpected source : %d:%d:%d",
				info->si_code, info->si_pid, info->si_uid);
		 
		return;
	}

	WARN("Authentication timer expired, aborting authentication");
	exit(EXIT_FAILURE);
}

static void 
sigpipe_action(int signum, siginfo_t *info, void *ctx __U)
{
	if (signum != SIGPIPE) {
		WARN("Unexpected signal %d, expected SIGPIPE", signum);
		return;
	}
	if (info->si_code != SI_USER || info->si_pid != getpid()) {
		WARN("Received SIGALRM from unexpected source : %d:%d:%d",
				info->si_code, info->si_pid, info->si_uid);
		return;
	}
	WARN("Client socket closed, aborting");
	exit(EXIT_FAILURE);
}


static int
install_sigactions(void)
{
	struct sigaction alrm_action = {
		.sa_sigaction = sigalrm_action,
		.sa_flags = SA_SIGINFO | SA_RESETHAND,
		.sa_restorer = NULL,
	};
	struct sigaction pipe_action = {
		.sa_sigaction = sigpipe_action,
		.sa_flags = SA_SIGINFO | SA_RESETHAND,
		.sa_restorer = NULL,
	};
	sigemptyset(&alrm_action.sa_mask);
	sigemptyset(&pipe_action.sa_mask);

	if (sigaction(SIGALRM, &alrm_action, NULL))
		return -1;
	if (sigaction(SIGPIPE, &pipe_action, NULL))
		return -1;
	if (signal(SIGCHLD, SIG_DFL) == SIG_ERR)
		return -1;
	return 0;
}


/*************************************************************/
/*                   'Self' type sockets                     */
/*************************************************************/

static char *
get_passwd(int s)
{
	char buf[PW_MAXLEN];
	ssize_t readlen;
	char *ret;

	readlen = clip_sock_read(s, buf, PW_MAXLEN, PW_READ_TIMEOUT, PW_TRIES);
	/* We don't deal with EINTR, etc... */
	if (readlen < 0) {
		WARN_ERRNO("could not read password");
		return NULL;
	}


	if (readlen == PW_MAXLEN) {
		WARN("password is too long\n");
		return NULL;
	}

	buf[readlen] = '\0';
	
	ret = strdup(buf);
	if (!ret) 
		WARN_ERRNO("could not copy password");

	return ret;
}

static int 
do_ack(int s, int retval)
{
	char c = (retval) ? 'N' : 'Y';

	if (write(s, &c, 1) != 1) {
		WARN_ERRNO("could not write ack");
		(void)close(s);
		return -1;
	}
	(void)close(s);
	return 0;
}

int 
check_user_self(int s, char *sockname __U)
{
	uid_t uid;
	gid_t gid;
	struct passwd *pwd;
	char *tmp;
	int ret;

	struct auth_info auth;

#ifdef USE_SYSLOG
	/* Reopen syslog in 'auth' facility */
	if (g_daemonized)
		openlog("pwcheckd", LOG_PID, LOG_AUTH);
#endif

	if (install_sigactions()) {
		WARN("could no install signal handlers");
		return -1;
	}

	if (alarm(PW_TIMEOUT)) 
		WARN("resetting previous timer");

	if (clip_getpeereid(s, &uid, &gid)) {
		WARN("could not get peer creds");
		return -1;
	}

#ifdef USE_TCB
	if (set_gid())
		return -1;

	if (setuid(uid)) {
		WARN_ERRNO("setuid failed");
		return -1;
	}
#endif

	pwd = getpwuid(uid);
	if (!pwd) {
		WARN_ERRNO("user not found");
		return -1;
	}

	tmp = strdup(pwd->pw_name);
	if (!tmp) {
		WARN("out of memory?\n");
		return -1;
	}
	auth.username = tmp;

	tmp = get_passwd(s);
	if (!tmp) {
		free(auth.username);
		return -1;
	}
	auth.passwd = tmp;

	ret = do_auth(&auth);

	if (ret)
		LOG("Authentication failure for %s", auth.username);
	else
		LOG("Successful authentication for %s", auth.username);

	free(auth.passwd);
	free(auth.username);

	ret |= do_ack(s, ret);
	return ret;
}
/*************************************************************/
/*                   'X11' type sockets                      */
/*************************************************************/

static int
run_x11_cmd(int fd, char *service, char *sockname)
{
	int ret;
	char tmp[sizeof(X11_SCRIPT)] = X11_SCRIPT;
	char *const argv[] = { tmp, service, sockname, NULL };
	char *envp[] = { NULL };

	(void)close(STDOUT_FILENO);
	
	if (dup2(fd, STDOUT_FILENO) == -1) {
		WARN_ERRNO("dup2() failed");
		return -1;
	}

	ret = -execve(argv[0], argv, envp);
	if (ret) 
		WARN("X11 command %s - %s failed: %d", X11_SCRIPT, 
							service, ret);
	return ret;
}

static int
read_child_auth(int fd, struct auth_info *auth) {
	char buf[2*PW_MAXLEN + 3];
	size_t len;
	ssize_t rret;
	char *endptr, *pwptr;

	rret = read(fd, buf, sizeof(buf));
	if (rret < 0) {
		WARN_ERRNO("read() failed");
		return -1;
	}
	len = rret;
	if (!len) {
		WARN("Unexpected EOF");
		return -1;
	}
	if (len == sizeof(buf)) {
		WARN("Password + username too long");
		return -1;
	}
	buf[len] = '\0';

	endptr = strstr(buf, X11_SEP);
	if (!endptr) {
		WARN("Wrong response from child");
		return -1;
	}
	if ((size_t)(endptr - buf) >= 
			(size_t)(len + 1 - sizeof(X11_SEP))) {
		WARN("Empty password");
		return -1;
	}
	pwptr = endptr + sizeof(X11_SEP) - 1;

	auth->username = strndup(buf, endptr - buf);
	if (!auth->username) {
		WARN("Out of memory for username");
		return -1;
	}
	auth->passwd = strdup(pwptr);
	if (!auth->passwd) {
		WARN("Out of memory for password");
		free(auth->username);
		return -1;
	}
	LOG("Current X11 user is %s", auth->username);

	return 0;
}

static int 
get_x11_auth(struct auth_info *auth, char *service, char *sockname)
{
	int fds[2];
	pid_t pid, wret;
	int status;

	if (pipe(fds)) {
		WARN_ERRNO("pipe() failed");
		return -1;
	}

	pid = fork();
	switch (pid) {
		case -1:
			WARN_ERRNO("fork() failed");
			return -1;
		case 0:
			(void)close(fds[0]);
			exit(run_x11_cmd(fds[1], service, sockname));
		default:
			(void)close(fds[1]);
			wret = waitpid(pid, &status, 0);
			if (wret < 0) {
				WARN_ERRNO("waitpid() X11 failed");
				return -1;
			}
			if (wret != pid) {
				WARN("waitpid() returned wrong child: %d != %d",
						wret, pid);
				return -1;
			}
			if (!WIFEXITED(status) || WEXITSTATUS(status)) {
				WARN("Abnormal child termination");
				return -1;
			}

			if (read_child_auth(fds[0], auth))
				return -1;
			return 0;
	}
			
	WARN("Not reached!");
	return -1;
}

int
check_user_x11(int s, char *sockname)
{
	int ret;

	struct auth_info auth;
	char service[PW_MAXLEN + 1];
	ssize_t rret;
	size_t len;

#ifdef USE_SYSLOG
	/* Reopen syslog in 'auth' facility */
	if (g_daemonized)
		openlog("pwcheckd", LOG_PID, LOG_AUTH);
#endif

	if (install_sigactions()) {
		WARN("could no install signal handlers");
		return -1;
	}

	if (alarm(X11_TIMEOUT)) 
		WARN("resetting previous timer");


	rret = read(s, service, sizeof(service));
	if (rret < 0) {
		WARN_ERRNO("read() service failed");
		return -1;
	}
	len = rret;
	if (!len) {
		WARN("Unexpected EOF while reading service name");
		return -1;
	}
	if (len == sizeof(service)) {
		WARN("Service name too long");
		return -1;
	}
	service[len] = '\0';

	if (get_x11_auth(&auth, service, sockname))
		return -1;

	ret = do_auth(&auth);

	if (ret)
		LOG("Authentication failure for %s", auth.username);
	else
		LOG("Successful authentication for %s", auth.username);

	free(auth.passwd);
	free(auth.username);

	ret |= do_ack(s, ret);
	return ret;
}


src/tester.c
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright © 2007-2018 ANSSI. All Rights Reserved.
#define _GNU_SOURCE
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>

#define PWLEN 64

#define WARN(fmt, args...) \
	fprintf(stderr, "%s: " fmt, __FUNCTION__, ##args)

#define _TO_STR(var) #var
#define TO_STR(var) _TO_STR(var)

#ifdef PWCHECKD_SOCKET
#define _PWCHECKD_SOCKET TO_STR(PWCHECKD_SOCKET)
#else
#define _PWCHECKD_SOCKET "/tmp/clip.sock"
#endif

static void
usage(const char *prog)
{
	printf("usage: %s <sock> <service|password>\n", prog);
}

static int
do_test(const char *sock, const char *service)
{
	char *ptr;
	struct sockaddr_un sau;
	ssize_t len;
	char c;
	int s, ret = -1;

	sau.sun_family = AF_UNIX;
	snprintf(sau.sun_path, sizeof(sau.sun_path), "%s", sock);
	
	s = socket(PF_UNIX, SOCK_STREAM, 0);
	if (s < 0) {
		perror("socket");
		return ret;
	}
	
	if (connect(s, (struct sockaddr *)&sau, sizeof(struct sockaddr_un)) < 0) {
		perror("connect");
		goto out;
	}
	
	len = write(s, service, strlen(service));
	if (len < 0) {
		perror("write");
		goto out;
	}	

	if (read(s, &c, 1) < 0) {
		perror("read");
		goto out;
	}

	if (c == 'Y')
		ret = 0;
out:
	close(s);
	return ret;
}

int 
main(int argc, char *argv[])
{
	if (argc < 3) {
		usage(basename(argv[0]));
		return EXIT_FAILURE;
	}

	return do_test(argv[1], argv[2]);
}
		
